### PR TITLE
Improve tests

### LINKED ISSUE #56: Refactor unit tests to properly be "unit tests"

### PR DESCRIPTION
**Pull request by gtupak**
Closes #56

- Refactor the `ArcxAnalyticsSdk.test.ts` file to proper unit tests
- Reach almost 100% coverage
- Should fix a bug causing the tests to randomly fail at run-time
- Move the mocha arguments to the `.mocharc.yaml` and `.mocharc-watch.yaml` files

image

### CODE REVIEW

File: test/ArcxAnalyticsSdk.test.ts [hunk 1]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
-            .getCall(1)
-            .calledWith(
-              PROD_URL_BACKEND,
-              '',
-              '/submit-event',
-              getAnalyticsData(CAUGHT_TRANSACTION_EVENT, { ...transactionParams, nonce }),
-            ),
-        ).to.be.true
+        expect(window.ethereum?.on).calledWithMatch('accountsChanged')
+      })
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:160]: What is the point of splitting it to two separate tests?

File: test/ArcxAnalyticsSdk.test.ts [hunk 2]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
       it('calls _reportCurrentWallet if trackWalletConnections is true', async () => {
-        // eslint-disable-next-line @typescript-eslint/no-extra-semi
-        ;(window as any).ethereum = new MockEthereum()
-
         const reportCurrentWalletStub = sinon.stub(
           ArcxAnalyticsSdk.prototype,
-          <any>'_reportCurrentWallet',
+          '_reportCurrentWallet' as any,
         )
         await ArcxAnalyticsSdk.init(TEST_API_KEY, { trackWalletConnections: true })
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:151]: All values should be true in `default_config` doesn't it?

File: test/ArcxAnalyticsSdk.test.ts [hunk 3]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
-              PROD_URL_BACKEND,
-              '',
-              '/submit-event',
-              getAnalyticsData(CAUGHT_TRANSACTION_EVENT, { ...transactionParams, nonce }),
-            ),
-        ).to.be.true
+        expect(window.ethereum?.on).calledWithMatch('accountsChanged')
+      })
+
+      it('calls _onAccountsChanged when accountsChanged is fied and trackWalletConnections is true', async () => {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:162]:
```suggestion
      it('calls _onAccountsChanged when accountsChanged is fired and trackWalletConnections is true', async () => {
```

File: test/ArcxAnalyticsSdk.test.ts [hunk 4]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
-    it('#page', async () => {
-      const pageAttributes = { url: 'page.test' }
-      const eventStub = sinon.stub(analyticsSdk, 'event')
+          getAnalyticsData('TEST_EVENT', testAttributes),
+        )
+      })
 
-      await analyticsSdk.page(pageAttributes)
+      it('supports nested attributes', async () => {
+        const pageAttributes = { layer1: { layer2: { layer3: { layer4: 'hello!' } } } }
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:200]: just `attributes` if it is local

File: test/ArcxAnalyticsSdk.test.ts [hunk 5]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
     })
 
-    it('#event', async () => {
-      await analyticsSdk.event('TEST_EVENT', TEST_ATTRIBUTES)
-      expect(
-        postRequestStub.calledOnceWith(
-          PROD_URL_BACKEND,
+    describe('#event', () => {
+      it('calls postRequest with the correct params', async () => {
+        const testAttributes = {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:185]: just `attributes` if it is local

File: test/ArcxAnalyticsSdk.test.ts [hunk 6]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
 
-    it('#connectWallet', async () => {
-      const attributes = { account: '0x12354', chain: '1' }
-      const eventStub = sinon.stub(analyticsSdk, 'event')
-
-      await analyticsSdk.connectWallet(attributes)
+    describe('#attribute', () => {
+      it('calls event() with the given attributes', async () => {
+        const eventStub = sinon.stub(analyticsSdk, 'event')
+        const testAttributes = {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:215]: just `attributes` if it is local

File: test/ArcxAnalyticsSdk.test.ts [hunk 7]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
+        expect(eventStub).calledOnceWithExactly(PAGE_EVENT, testAttributes)
+      })
     })
 
-    it('#referrer', async () => {
-      const eventStub = sinon.stub(analyticsSdk, 'event')
+    describe('#connectWallet', () => {
+      it('calls event() with the given attributes', async () => {
+        const eventStub = sinon.stub(analyticsSdk, 'event')
+        const testAttributes = {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:239]: just `attributes` if it is local

File: test/ArcxAnalyticsSdk.test.ts [hunk 8]
```diff
@@ -1,70 +1,84 @@
...
+    postRequestStub.resetHistory()
+    localStorage.clear()
+    sessionStorage.clear()
+  })
 
-    afterEach(() => sinon.restore())
+  afterEach(sinon.restore)
 
+  describe('Public Functions', () => {
     describe('#init', () => {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:63]: Initialisation should be moved outside of 'Public Functions'. In my opinion it is like contstructor, not Public function. Also what is the point of splitting functions into `Private functions` and `Public Functions` if you introduced naming convention and we already have difference in naming

File: test/ArcxAnalyticsSdk.test.ts [hunk 9]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
+      })
     })
 
-    it('#connectWallet', async () => {
-      const attributes = { account: '0x12354', chain: '1' }
-      const eventStub = sinon.stub(analyticsSdk, 'event')
-
-      await analyticsSdk.connectWallet(attributes)
+    describe('#attribute', () => {
+      it('calls event() with the given attributes', async () => {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:213]: You use `it('calls event() with the given attributes', async () => {` in every test. In my opinion it doesn't provide any additional information. So we can just get rid of this it and leave  `it('#[functionName', () => {`.
if you think that it provides additional sense, it is equal to situation :
```ts
describe('calls event() with the given attributes', async () => {
  it('#function1', () => {...}
  ....
   it('#functionN', () => {...}
}
```

File: test/ArcxAnalyticsSdk.test.ts [hunk 10]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
 
-    it('super nested attributes are supported', async () => {
-      const pageAttributes = { layer1: { layer2: { layer3: { layer4: 'hello!' } } } }
-      const eventStub = sinon.stub(analyticsSdk, 'event')
+      it('emits a FIRST_PAGE_VISIT event with url if trackPages is true', () => {
+        analyticsSdk['sdkConfig'].trackPages = true
 
-      await analyticsSdk.event('TEST_EVENT', pageAttributes)
+        analyticsSdk['_trackFirstPageVisit']()
+        expect(eventStub).calledOnceWithExactly(FIRST_PAGE_VISIT, { url: TEST_JSDOM_URL })
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:296]: we can just check if `_trackFirstPageVisit` is called. But event is good as well

File: test/ArcxAnalyticsSdk.test.ts [hunk 11]
```diff
@@ -89,290 +110,460 @@ describe('(unit) ArcxAnalyticsSdk', () => {
...
+
+        analyticsSdk['sdkConfig'].trackPages = false
+        analyticsSdk['sdkConfig'].trackReferrer = false
+        analyticsSdk['sdkConfig'].trackUTM = false
+      })
     })
 
-    afterEach(sinon.restore)
+    describe('#_trackPagesChange', () => {
+      it('registers an on-click event listener', () => {
```
Comment (xitronix) [inline test/ArcxAnalyticsSdk.test.ts:348]: `_trackPagesChange: registers an on-click event listener` 
I don't get the idea to create context for single test

**Review by xitronix — CHANGES_REQUESTED** [review]

**Review by xitronix — COMMENTED** [review]
Do wee still need MockEthereum used?

File: test/fixture.ts [hunk 12]
```diff
@@ -1,3 +1,10 @@
...
+export const TEST_UTM_SOURCE = 'facebook'
+export const TEST_UTM_CAMPAIGN = 'ad-camp'
+export const TEST_UTM_MEDIUM = 'cpc'
+export const TEST_REFERRER = 'https://arcx.money/'
+export const TEST_JSDOM_URL = `https://example.com/?utm_source=${TEST_UTM_SOURCE}&utm_medium=${TEST_UTM_MEDIUM}&utm_campaign=${TEST_UTM_CAMPAIGN}`
+
+export const TEST_API_KEY = '01234'
+export const TEST_ACCOUNT = '0x1234567890123456789012345678901234567890'
 export const TEST_CHAIN_ID = '1'
 export const TEST_IDENTITY = 'test-identity'
```
Comment (gtupak) [inline test/fixture.ts:10]: Take out the constants and keep the ones that are actually used in fixture

File: test/ArcxAnalyticsSdk.test.ts [hunk 13]
```diff
@@ -1,70 +1,84 @@
...
-import { TEST_ADDRESS, TEST_CHAIN_ID, TEST_IDENTITY } from './fixture'
+} from './fixture'
+import { MetaMaskInpageProvider } from '@metamask/providers'
 import { MockEthereum } from './MockEthereum'
-import { ArcxAnalyticsSdk } from '../src'
-
-const PROD_URL_BACKEND = DEFAULT_SDK_CONFIG.url // Backwards compatability
 
 const ALL_FALSE_CONFIG: SdkConfig = {
+  ...DEFAULT_SDK_CONFIG,
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:34]: Will change the type to `Partial<SdkConfig>`

File: test/ArcxAnalyticsSdk.test.ts (hunk 9 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:213]: If you have a describe with one `it`, get rid of `describe` and only use `it`

File: test/ArcxAnalyticsProvider.test.tsx [hunk 14]
```diff
@@ -75,14 +77,23 @@ const ChildTest = () => {
...
 }
 
-describe('ArcxAnalyticxProvider', () => {
+describe('(int) ArcxAnalyticxProvider', () => {
   let postRequestStub: sinon.SinonStub
   let screen: RenderResult
 
   beforeEach(async () => {
+    const ethereumMock = sinon.createStubInstance(MetaMaskInpageProvider)
+    window.ethereum = ethereumMock
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsProvider.test.tsx:86]: done

File: test/ArcxAnalyticsSdk.test.ts [hunk 15]
```diff
@@ -1,70 +1,84 @@
...
-      analyticsSdk = await ArcxAnalyticsSdk.init(TEST_API_KEY, ALL_FALSE_CONFIG)
-      postRequestStub.resetHistory()
-      sessionStorage.clear()
-    })
+  let analyticsSdk: ArcxAnalyticsSdk
+  let postRequestStub: sinon.SinonStub
+
+  beforeEach(async () => {
+    const ethereumMock = sinon.createStubInstance(MetaMaskInpageProvider)
+    window.ethereum = ethereumMock
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:50]: done

File: src/ArcxAnalyticsSdk.ts [hunk 16]
```diff
@@ -165,7 +165,7 @@ export class ArcxAnalyticsSdk {
     const chainIdHex = await window.ethereum.request<string>({ method: 'eth_chainId' })
     // Because we're connected, the chainId cannot be null
     if (!chainIdHex) {
-      throw new Error('ArcxAnalyticsSdk::_getCurrentChainId: chainIdHex is:' + chainIdHex)
+      throw new Error('ArcxAnalyticsSdk::_getCurrentChainId: chainIdHex is: ' + chainIdHex)
```
[the remark this answers is not in the document]
Comment (gtupak) [reply src/ArcxAnalyticsSdk.ts:168]: done

File: test/ArcxAnalyticsSdk.test.ts [hunk 17]
```diff
@@ -1,70 +1,84 @@
...
+    postRequestStub = sinon.stub(postRequestModule, 'postRequest').resolves(TEST_IDENTITY)
+
+    analyticsSdk = await ArcxAnalyticsSdk.init(TEST_API_KEY, ALL_FALSE_CONFIG)
+
+    postRequestStub.resetHistory()
+    localStorage.clear()
+    sessionStorage.clear()
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:57]: changed to after

File: test/ArcxAnalyticsSdk.test.ts [hunk 18]
```diff
@@ -1,70 +1,84 @@
...
 
-    afterEach(() => sinon.restore())
+  afterEach(sinon.restore)
 
+  describe('Public Functions', () => {
     describe('#init', () => {
-      it('posts identify', async () => {
+      it('does not get the identity from localStorage if `cacheIdentity` is false', async () => {
+        const localStorageStub = sinon.stub(localStorage, 'getItem')
+        await ArcxAnalyticsSdk.init(TEST_API_KEY, { ...ALL_FALSE_CONFIG, cacheIdentity: false })
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:66]: updated

File: test/ArcxAnalyticsSdk.test.ts [hunk 19]
```diff
@@ -1,70 +1,84 @@
...
+        expect(localStorageStub).to.not.have.been.called
+      })
+
+      it('sets the identity in localStorage if `cacheIdentity` is true', async () => {
+        expect(localStorage.getItem(IDENTITY_KEY)).to.be.null
+        await ArcxAnalyticsSdk.init(TEST_API_KEY, { ...ALL_FALSE_CONFIG, cacheIdentity: true })
+        expect(localStorage.getItem(IDENTITY_KEY)).to.equal(TEST_IDENTITY)
+      })
```
[the remark this answers is not in the document]
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:74]: done

File: test/ArcxAnalyticsSdk.test.ts (hunk 1 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:160]: Merged into 1

File: test/ArcxAnalyticsSdk.test.ts (hunk 4 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:200]: done

File: test/ArcxAnalyticsSdk.test.ts (hunk 5 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:185]: done

File: test/ArcxAnalyticsSdk.test.ts (hunk 6 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:215]: done

File: test/ArcxAnalyticsSdk.test.ts (hunk 7 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:239]: done

File: test/ArcxAnalyticsSdk.test.ts (hunk 8 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:63]: refactored

File: test/ArcxAnalyticsSdk.test.ts (hunk 9 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:213]: I considered, it, but it looks cleaner in the console when you run the tests: 
image
If we don't use `describe` it makes it harder to differentiate between function tests and other tests

File: test/ArcxAnalyticsSdk.test.ts (hunk 11 above)
Comment (gtupak) [reply test/ArcxAnalyticsSdk.test.ts:348]: as described above

File: test/fixture.ts (hunk 12 above)
Comment (gtupak) [reply test/fixture.ts:10]: renamed to constants

File: test/ArcxAnalyticsSdk.test.ts (hunk 9 above)
Comment (xitronix) [reply test/ArcxAnalyticsSdk.test.ts:194]: ok

**Review by xitronix — APPROVED** [review]

**Pull request merged**